apimachinery/reflect: add DeepEqualWithOptions to support option IgnoreUnexportedFields

tlwr opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

relates to kubernetes/kubernetes#124154

this change changes the call signature of the wrapped method deepValueEqual, leaving the wrapped behaviour the same

we also add a new function PublicDeepEqual which does not panic when encountering an unexported field

for use cases like the controller-runtime, we are fetching info from the k8s API and then comparing it via reflection with an in-memory representation of what the state of the entity should be. If there are differences then an update needs to happen (ie in the CreateOrUpdate fn)

there are CRD users like Istio and Cilium which use structs with unexported fields (eg the structs are used via grpc/proto and have some protocol level state bookkeeping), it should be possible to use these in a controller without having to redefine the structs, hence ignoring unexported fields - especially since unexported fields are not going to get serialized when sending/retrieving to/from the k8s API

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

Relates to kubernetes/kubernetes#124154

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
New helper PublicDeepEqual added to apimachinery reflection package
```

tlwr committed:
apimachinery/reflect: add fn DeepEqualWithOptions

relates to kubernetes/kubernetes#124154

this change changes the call signature of the wrapped method
deepValueEqual, leaving the wrapped behaviour the same

we also add a new function DeepEqualWithOptions which allows
customizable behaviour of DeepEqual, so we do not need additional
combinatorial functions. The new flag allows us to ignore unexported
fields when checking equality, instead of panicking

for use cases like the controller-runtime, we are fetching info from the
k8s API and then comparing it via reflection with an in-memory
representation of what the state of the entity should be. If there are
differences then an update needs to happen (ie in the CreateOrUpdate fn)

there are CRD users like Istio and Cilium which use structs with
unexported fields (eg the structs are used via grpc/proto and have some
protocol level state bookkeeping), it should be possible to use these in
a controller without having to redefine the structs, hence ignoring
unexported fields - especially since unexported fields are not going to
get serialized when sending/retrieving to/from the k8s API

Signed-off-by: toby lorne <toby@toby.codes>

fedebongio commented:
/assign @jpbetz @sttts @Jefftree 

Thank you for the PR @tlwr. This seems to be a delicate change, and requires a bit more consideration. I've tagged a couple of experts. The consensus during the triage meeting was that we needed more context on this change and some more discussion. Maybe come to the SIG API Machienry meeting? or on this same bug, whatever works. Thanks!

tlwr commented:
I have added the SIG meeting to my calendar and will attempt to join next Wednesday, but happy to discuss or answer any questions in this PR, or in the linked issue

Additional context on the subsequent change (assuming the API addition as-is)
https://github.com/kubernetes-sigs/controller-runtime/blob/e6c3d139d2b6c286b1dbba6b6a95919159cfe655/pkg/controller/controllerutil/controllerutil.go#L312-L314

```
+	// ignore unexported fields because the existing is populated from API state
+	// and will not have these unexported members set
+	if equality.Semantic.PublicDeepEqual(existing, obj) {
-	if equality.Semantic.DeepEqual(existing, obj) {
 		return OperationResultNone, nil
 	}
```

jpbetz commented on staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:
```diff
@@ -270,14 +274,18 @@ func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool,
+	return e.deepEqual(a1, a2, false, false)
+}
+
+func (e Equalities) PublicDeepEqual(a1, a2 interface{}) bool {
+	return e.deepEqual(a1, a2, true, true)
```
I don't know if a function name per combination of options is the right way to continue to expand here..  Will we need a `PublicDeepEqualWithNilDifferentFromEmpty` later?  What if we add another independently settable option?

tlwr replied:
Happy to have a different API than the one currently suggested, but introducing a new function without changing the call signature of `equality.Semantic.DeepEqual` is less intrusive on the [existing consumers](https://github.com/search?q=equality.Semantic.DeepEqual+k8s.io%2Fapimachinery%2Fpkg%2Fapi%2Fequality+language%3AGo&type=code&l=Go) of which there are quite a few

jpbetz replied:
We can't change the existing functions.  But if we decide to add a new function, how about a name like `DeepCopyWithOptions`?

tlwr replied:
Yes that sounds reasonable, I propose:

We wrap the current helper method:
```
func (e Equalities) deepEqual(a1, a2 interface{}, equateNilAndEmpty, ignoreUnexportedFields bool) bool {
```
with a new method with the name you suggest: `DeepEqualWithOptions` (I assume you meant equal instead of copy)

but I do not like having a list of bools as arguments (if future options are added later), we could use a single argument with a iota of flags?, so the following type and signature:

```
type DeepEqualOpts uint8

const (
  DeepEqualEquateNilAndEmpty DeepEqualOpts = 1 << iota
  DeepEqualIgnoreUnexportedFields
)

func (e Equalities) DeepEqualWithOptions(a1, a2 interface{}, flags DeepEqualOpts) bool { 
```

and adding a new option will not break the API for existing consumers

WDYT?

jpbetz replied:
How about making the options a struct? We then just require that any change to the struct is backward compatible.  Initially it's flags that default to false, but we're not limited to flag options in the future

tlwr replied:
Yes, sounds good, I have updated the PR with these changes

jpbetz commented on staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal_test.go:
```diff
@@ -98,6 +98,40 @@ func TestEqualities(t *testing.T) {
+		y int
+	}
+
+	variantA := Foo{1, 2}
```
I'm not sure different int values help with coverage.  How about testing different compositions?  Fields containing structs?  Embeddedd structs? Arrays of structs, maps with struct values?  Unexported structs?  ...

jpbetz commented on staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal_test.go:
```diff
@@ -98,6 +98,40 @@ func TestEqualities(t *testing.T) {
+		IgnoreUnexportedFields: true,
+	}
+
+	if e.DeepEqualWithOptions(variantA, variantB, opts) {
```
Table driven tests would be nice here

https://go.dev/wiki/TableDrivenTests

tlwr committed:
apimachinery: table tests DeepEqualWithOptions

Signed-off-by: toby lorne <toby@toby.codes>

jpbetz commented:
Test structure looks good.  Can we add privates types too?  For example, what happens if an exported field has an unexported type (go allows it)?

What other traversals should be covered?  How about a pointer to a unexported type?  Let's make sure we have solid coverage of all the ways the traversal code can navigate types.

jpbetz commented:
Another example of a corner case: Unexported embedded type with exported field in an exported type.

tlwr committed:
apimachinery: add edge cases DeepEqualWithOptions

Signed-off-by: toby lorne <toby@toby.codes>

tlwr commented:
Added additional edge cases 👍

tlwr committed:
apimachinery: more edge cases DeepEqualWithOptions

Signed-off-by: toby lorne <toby@toby.codes>

tlwr commented:
Added some more cases both for exported/unexported and for some hard types covered by reflection statements

Jefftree commented:
/triage accepted

tlwr commented:
/remove-lifecycle stale

sttts commented:
/ok-to-test

tlwr commented:
/remove-lifecycle stale

strideynet commented:
Hey folks, is there anything we can do to help move this along? From my understanding this'll unblock https://github.com/kubernetes/kubernetes/issues/124154 which we've been stuck on for a while.

tlwr commented:
@strideynet not sure if/when the k8s apimachinery maintainers will merge this

I am currently working around the issue like so

(in a pkg's init)

```
    // istio structs contain unexported fields used for grpc which cause the
    // controller-runtime DeepEqual function to panic.
    // the suggestion from the kubernetes developers is to add a custom equality function.
    // to do this we can use the google cmp library with ignoring the grpc fields
    ignoreGRPCFields := gocmp.FilterPath(func(path gocmp.Path) bool {
        fname := path.Last().String()
        return fname == ".state" || fname == ".sizeCache" || fname == ".unknownFields"
    }, gocmp.Ignore())

    // linter will complain but we need to pass the concrete type, otherwise no match in DeepEqual
    equality.Semantic.AddFunc(func(a, b istionetworkingv1beta1.Sidecar) bool {
        return gocmp.Equal(&a, &b, ignoreGRPCFields)
    })

    // linter will complain but we need to pass the concrete type, otherwise no match in DeepEqual
    equality.Semantic.AddFunc(func(a, b istiov1alpha1.IstioStatus) bool {
        return gocmp.Equal(&a, &b, ignoreGRPCFields)
    })
 ```

but the downside of this is that each individual struct needs to be added, so it is quite fragile
I recommend unit tests that test the serialization/deserialization to avoid runtime errors

tlwr commented:
@jpbetz, to clarify - this MR as is  allows users of this package to avoid needing a dependency on `google/go-cmp`  package or similar comparator, so the above would be:

```
    // linter will complain but we need to pass the concrete type, otherwise no match in DeepEqual
    equality.Semantic.AddFunc(func(a, b istionetworkingv1beta1.Sidecar) bool {
        return equality.Semantic.DeepEqual(&a, &b, equality.DeepEqualOptions{IgnoreUnexportedFields: true})
    })

    // linter will complain but we need to pass the concrete type, otherwise no match in DeepEqual
    equality.Semantic.AddFunc(func(a, b istiov1alpha1.IstioStatus) bool {
        return equality.Semantic.DeepEqual(&a, &b, equality.DeepEqualOptions{IgnoreUnexportedFields: true})
    })
```

(ie use equality package instead of go-cmp)

One still needs to:

* do the testing, to ensure forward compatibility
* specify the function for each type individually (so testing extra important), because this MR doesn't change the default comparison behaviour

So I think this MR is still useful, because it gives people who are using this package ability to interoperate nicely with structs using private fields (which is very common for grpc and similar use-cases), without pulling in another dependency

tlwr commented:
/remove-lifecycle stale

jpbetz reviewed (commented):
Let's retitle this to reflect that `IgnoreUnexportedFields` is being introduced and a `DeepEqualWithOptions` function also being added.

jpbetz reviewed (commented):
In addition to https://github.com/kubernetes/kubernetes/pull/126998#issuecomment-3928517619 here's some line-by-line review.

tlwr committed:
add DeepEqualWithOptions test for time.Time

Signed-off-by: toby lorne <toby@toby.codes>

tlwr committed:
use t.Run for TestPublicEqualities

Signed-off-by: toby lorne <toby@toby.codes>

tlwr committed:
DeepEqualWithOptions has correct defaults and test

as pointed out in PR review, default value would have unexpected
behaviour vs normal deep equal operation, renaming the option to provide
correct default value

Signed-off-by: toby lorne <toby@toby.codes>

tlwr commented:
/remove-lifecycle stale
